core/common/src/main/java/alluxio/exception/DirectoryNotInSyncException.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.exception;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * Exception for situations where an operation fails because a directory is not in sync with the
+ * UFS. An example of this would be deleting a nonempty directory which has UFS contents not loaded
+ * into Alluxio.
+ */
+@ThreadSafe
+public class DirectoryNotInSyncException extends AlluxioException {
```

Comment:
I agree that it's important to handle it differently, e.g. not retrying. All of our server APIs classify exceptions into one of 13 categories defined in `Status.java`. Only the `Unavailable` status is retried. Unrecognized `IOExceptions` are translated to the `Unavailable` status. However, `FailedPreconditionException` is understood by our RPC layer and will be translated to the `FailedPrecondition` status on the wire, so it won't be retried.